cosmetics-web/app/models/component.rb

```diff
@@ -135,16 +135,28 @@ def display_root_category
     get_category_name(root_category)
   end
 
+  # This should read formulation file required
+  # TODO: find where it is used
   def formulation_required?
     if range?
       !formulation_file.attached? && range_formulas&.empty?
     elsif exact?
       !formulation_file.attached? && exact_formulas&.empty?
+    elsif predefined? && contains_poisonous_ingredients
+      !formulation_file.attached?
     else
       false
     end
   end
```

Comment:
Suggestion: We can simplify the method by using a guard clause:

```suggestion
  def formulation_required?
    return false if formulation_file.attached?

    (range && range_formulas&.empty?) ||
      (exact? && exact_formulas&.empty?) ||
      (predefined? && contains_poisonous_ingredients)
  end
```